extensions/2.0/Vendor/FB_materials_lightmap/README.md

```diff
@@ -0,0 +1,76 @@
+# FB\_materials\_lightmap
+
+## Contributors
+
+* Alfred Young, Facebook, [@idorurez](https://github.com/idorurez)
+* Pär Winzell, Facebook, [@zellski](https://twitter.com/zellski)
+
+## Status
+
+Draft
+
+## Dependencies
+
+Written against the glTF 2.0 spec.
+
+## Overview
+
+This extension adds an additional annotation to a material, providing a means to introduce a lightmap to be applied in the same scope.
+
+Some proven real-world applications for this information:
+1. Having multiple variations of lighting scenarios without having to bake lighting into a single, original base color/texture.
+2. Specify a separate UV channel (texCoord), to vary the new texture texel resolution with respect to the texture it's being applied to.
+3. Modulate the intensity of a lightmap, dimming or increasing it's effect.
+
+## Specifying Material extension
+
+The data is defined as an optional extension to a glTF `materials`, by adding an `extensions` property and defining a `FB_materials_lightmap` property with a `lightmapTexture` node inside it:
+
+```javascript
+"materials" : [
+  	{
+  		"name" : "material_test",
+  		"pbrMetallicRoughness" : {
+  			"baseColorTexture" : {
+  				"index" : 0,
+  				"texCoord" : 0
+  			}
+  		},
+  		"extensions" : {
+  			"FB_materials_lightmap" : {
+  				"lightmapTexture" : {
+  					"index" : 1,
+  					"texCoord" : 1,    
+  				},
+				"lightmapFactor" : [
+  						1,
+  						1,
+  						1
+  				]
+  			}
+  		}
+  	}
+  ],
+```
+
+## Sample Implementation
+
+In it's simplemest form, Lightmap implementation predominantely involes a base, unlit texture or color in a material, and the supplied lightmap texture is sampled and multiplied against that unlit texture/color.
+
+```javascript
+    lowp vec4 lightmapColor = texture2D(LightmapTexture, oLightmapTexCoord);
+    color.rgb *= (LightmapFactor.rgb * lightmapColor.rgb);
```

Comment:
For threejs at least, this would be `+=` rather than `*=`... do we have other references here?

Reply:
Yes, this is a draft for a vendor extension and I'd like to merge it as-is, and it's already been implemented. **However**, I'm soliciting a broader range of practices in the event I'm completely missing an important component.  Furthermore, my limited research along with internal discussion hasn't revealed anything beyond a simple multiplicative application for a very basic lightmaps. I am aware of the additive application, but I am under the assumption that such an application implies that there still needs to be some level of

Reply:
> Furthermore, my limited research (I'll take a look at the Unity link above also) along with internal discussion hasn't revealed anything beyond a simple multiplicative application for a very basic lightmaps.

Ok thanks. I need to learn more about this myself, too. Everything I can find describes multiplicative lightmaps, so perhaps threejs is unusual here. We can dismiss the additive use if there are no other implementations.

I'm not sure I understand the advantage of a multiplicative lightmap over an ambient occlusion map – it's RGB, true, but can still only make the surface darker the base color?

> I am under the assumption that such an application implies that there still needs to be some level of lighting in the color/texture it is applied to...

If this extension is intended to replace other lighting in the scene, it could arguably require `KHR_materials_unlit` to also be present. But I would prefer to define both cases – both unlit and PBR application of the lightmap.

Reply:
> Ok thanks. I need to learn more about this myself, too. Everything I can find describes multiplicative lightmaps, so perhaps threejs is unusual here. We can dismiss the additive use if there are no other implementations.

Without revealing too much about what we have under our hood, we also have an unpublished spec that uses an additive application. After some consultation, we realized we needed to define a more broader, simple approach with a multiplicative application.

> I'm not sure I understand the advantage of a multiplicative lightmap over an ambient occlusion map – it's RGB, true, but can still only darken the base color?

The multiplicative aspect, in general, is the same. The main difference is that the source itself is indirect (occlusion) vs direct. I could easily see a situation where we would want to brighten the environment with the lightmap.

> If this extension is intended to replace other lighting in the scene, it could arguably require `KHR_materials_unlit` to also be present. But I would prefer to define both cases – both unlit and PBR application of the lightmap.

Ah good point. Again, without revealing too much, I have been working in an environment that operates under the assumption that the (PBR) material will always be unlit. What would you suggest we do in this case?

Reply:
> For threejs at least, this would be += rather than *=... do we have other references here?

Light maps, if you are doing them correctly, should be additive with all other light sources.

The reason is pretty simple.  Light in general is additive and a light map is not exclusive but rather just one more light source in a scene.  This way you can combine dynamic lights with precomputed light maps.

Reply:
> The multiplicative aspect, in general, is the same. The main difference is that the source itself is indirect (occlusion) vs direct. I could easily see a situation where we would want to brighten the environment with the lightmap.

>  source itself is indirect (occlusion) vs direct

A light source can not be occlusion and also a light source.  There are direct light sources and indirect light sources, and there are also occlusions, but all three of these things are separate.

I think you have an occlusion map and are calling it a light map.

> Yes, this is a draft for a vendor extension and I'd like to merge it as-is, and it's already been implemented.

If we basically want to say that this is a confused feature and that is okay because it is just a vendor extension, then sure, but then way standardize it at all?

Reply:
> I think you have an occlusion map and are calling it a light map.

That's a fair assessment.
 
> If we basically want to say that this is a confused feature and that is okay because it is just a vendor extension, then sure, but then why standardize it at all?

I'm not asking for a full EXT_, multiparty, approved extension. Does a vendor extension need to be standardized?

Reply:
> Does a vendor extension need to be standardized?

It does not. Very often contributors opening a "vendor" PR are also proposing standardization, or don't understand the difference, so we do tend to review liberally regardless of the prefix. :)